test/integration/consul-container/libs/cluster/cluster.go

```diff
@@ -121,6 +123,60 @@ func (c *Cluster) Remove(n libagent.Agent) error {
 	return nil
 }
 
+// StandardUpgrade upgrades a running consul cluster following the steps from
+//
+//	https://developer.hashicorp.com/consul/docs/upgrading#standard-upgrades
+//
+// - takes a snapshot
+// - terminate and rejoin the new version of consul
+func (c *Cluster) StandardUpgrade(t *testing.T, ctx context.Context, targetVersion string) error {
+	execCode, err := c.Agents[0].Exec(context.Background(), []string{"consul", "snapshot", "save", "backup.snap"})
+	if execCode != 0 {
+		return fmt.Errorf("error taking snapshot of the cluster, returned code %d", execCode)
+	}
+	if err != nil {
+		return err
+	}
+
+	// verify only the leader can take a snapshot
+	snapshotCount := 0
+	for _, agent := range c.Agents {
+		files, err := ioutil.ReadDir(filepath.Join(agent.DataDir(), "raft", "snapshots"))
+		if err != nil {
+			return err
+		}
+		if len(files) >= 1 {
+			snapshotCount++
+		}
+	}
+
+	if snapshotCount != 1 {
+		return fmt.Errorf("only leader agent can have a snapshot file, got %d", snapshotCount)
+	}
+
+	// Upgrade individual agent to the target version
+	client := c.Agents[0].GetClient()
+	for _, agent := range c.Agents {
+		agent.Terminate()
```

Comment:
Do we need to run `agent.Exec([]string{"consul", "leave"})` before terminate the container?

Reply:
Honestly, I'm not 100% sure here, but I know this will de-register services associated with agents. You might end up with an empty catalog. 

Maybe a more realistic scenario is that we should use `leave` and agents should have some service registrations they bootstrap in their config?